src/coreWebVitals.ts

```diff
@@ -0,0 +1,127 @@
+
+export type CoreWebVitalsPayload = {
+	page_view_id: string | null;
+	browser_id: string | null;
+	fid: null | number;
+	cls: null | number;
+	lcp: null | number;
+	fcp: null | number;
+	ttfb: null | number;
+};
+
+const coreWebVitalsPayload: CoreWebVitalsPayload = {
+	browser_id: null,
+	page_view_id: null,
+	fid: null,
+	cls: null,
+	lcp: null,
+	fcp: null,
+	ttfb: null,
+};
+
+let shouldSendMetrics = false;
+export const forceSendMetrics = (): void => {
+	shouldSendMetrics = true;
+};
+
+const roundWithDecimals = (value: number, precision = 6): number => {
+	const power = Math.pow(10, precision);
+	return Math.round(value * power) / power;
+};
+
+const sendData = (isDev = true): boolean => {
+	if (!shouldSendMetrics || coreWebVitalsPayload.fcp === null) return false;
+
+	const endpoint = isDev
+		? 'https://performance-events.code.dev-guardianapis.com/core-web-vitals'
+		: 'https://performance-events.guardianapis.com/core-web-vitals';
+
+	return navigator.sendBeacon(endpoint, JSON.stringify(coreWebVitalsPayload));
+};
+
+const onReport: ReportHandler = (metric) => {
+	switch (metric.name) {
+		case 'FCP':
+			// Browser support: Chromium, Firefox, Safari Technology Preview
+			coreWebVitalsPayload.fcp = roundWithDecimals(metric.value);
+			break;
+		case 'CLS':
+			// Browser support: Chromium,
+			coreWebVitalsPayload.cls = roundWithDecimals(metric.value);
+			break;
+		case 'LCP':
+			// Browser support: Chromium
+			coreWebVitalsPayload.lcp = roundWithDecimals(metric.value);
+			break;
+		case 'FID':
+			// Browser support: Chromium, Firefox, Safari, Internet Explorer (with the polyfill)
+			coreWebVitalsPayload.fid = roundWithDecimals(metric.value);
+			break;
+		case 'TTFB':
+			// Browser support: Chromium, Firefox, Safari, Internet Explorer
+			coreWebVitalsPayload.ttfb = roundWithDecimals(metric.value);
+			break;
+	}
+};
+
+/**
+ * Initialise sending Core Web Vitals metrics to a logging endpoint.
+ *
+ * @param init - the initialisation options
+ * @param init.isDev - Whether to log to CODE or PROD endpoints
+ * @param metricsSentCallback - Optional callback, triggered after metrics are queued for sending
+ */
+export const initCoreWebVitals = (
+	{
+		browserId,
+		pageViewId,
+		forceSendMetrics = false,
+		isDev,
+	}: {
+		browserId: string;
+		pageViewId: string;
+		isDev?: boolean;
+		forceSendMetrics?: boolean;
+	},
+	metricsSentCallback?: (queued?: boolean) => void,
+): void => {
+	coreWebVitalsPayload.browser_id = browserId;
+	coreWebVitalsPayload.page_view_id = pageViewId;
+
+	// By default, sample 1% of page views
+	const pageViewInSample = Math.random() < 1 / 100;
+	// Unless we are forcing sending metrics for this page view
+	// via initialisation or calling forceSendMetrics()
+	if (forceSendMetrics || pageViewInSample) shouldSendMetrics = true;
+	// Or using a specific hash
+	if (window.location.hash === '#forceSendMetrics') shouldSendMetrics = true;
```

Comment:
Well done, this code looks really simple but I can see how it could have been confusing but you've avoided that and done it really well 👍

Reply:
Thanks for the kind words. I think your input has made it even nicer. 🌷